pkg/reconciler/buildrun/resources/service_accounts.go

```diff
@@ -38,70 +39,112 @@ func IsGeneratedServiceAccountUsed(buildRun *buildv1alpha1.BuildRun) bool {
-		// succeeded to create the service account but failed to update the build run that references it
-		ctxlog.Info(ctx, "create or update serviceAccount for BuildRun", namespace, buildRun.Namespace, name, serviceAccountName, "BuildRun", buildRun.Name)
-		op, err := controllerutil.CreateOrUpdate(ctx, client, serviceAccount, func() error {
-			serviceAccount.SetLabels(map[string]string{buildv1alpha1.LabelBuildRun: buildRun.Name})
+	// generate or retrieve an existing autogenerated sa
+	if IsGeneratedServiceAccountUsed(buildRun) {
+		return GenerateSA(ctx, client, build, buildRun)
+	}
 
-			ownerReference := metav1.NewControllerRef(buildRun, buildv1alpha1.SchemeGroupVersion.WithKind("BuildRun"))
-			serviceAccount.SetOwnerReferences([]metav1.OwnerReference{*ownerReference})
-			serviceAccount.AutomountServiceAccountToken = &mountSAToken
-			ApplyCredentials(ctx, build, serviceAccount)
+	serviceAccount := &corev1.ServiceAccount{}
 
-			return nil
-		})
+	if buildRun.Spec.ServiceAccount != nil && buildRun.Spec.ServiceAccount.Name != nil {
+		serviceAccountName := *buildRun.Spec.ServiceAccount.Name
+		err := client.Get(ctx, types.NamespacedName{Name: serviceAccountName, Namespace: buildRun.Namespace}, serviceAccount)
+		// here we might need to update Status Conditions and Fail the BR
 		if err != nil {
+			if apierrors.IsNotFound(err) {
+				if updateErr := UpdateConditionWithFalseStatus(ctx, client, buildRun, fmt.Sprintf("service account %s not found", serviceAccountName), ConditionServiceAccountNotFound); updateErr != nil {
+					return nil, HandleError("failed to retrieve service account", err, updateErr)
+				}
+			}
+
 			return nil, err
 		}
-		ctxlog.Debug(ctx, "automatic generation of service account", namespace, serviceAccount.Namespace, name, serviceAccount.Name, "Operation", op)
 	} else {
-		// If ServiceAccount or the name of ServiceAccount in buildRun is nil, use pipeline serviceaccount
-		if buildRun.Spec.ServiceAccount == nil || buildRun.Spec.ServiceAccount.Name == nil {
-			serviceAccountName := PipelineServiceAccountName
-			err := client.Get(ctx, types.NamespacedName{Name: serviceAccountName, Namespace: buildRun.Namespace}, serviceAccount)
-			if err != nil && !apierrors.IsNotFound(err) {
-				return nil, err
-			} else if apierrors.IsNotFound(err) {
-				serviceAccountName = DefaultServiceAccountName
-				ctxlog.Info(ctx, "falling back to default serviceAccount", namespace, buildRun.Namespace)
-				err = client.Get(ctx, types.NamespacedName{Name: serviceAccountName, Namespace: buildRun.Namespace}, serviceAccount)
-				if err != nil {
-					return nil, err
-				}
-			}
-		} else {
-			serviceAccountName := *buildRun.Spec.ServiceAccount.Name
-			err := client.Get(ctx, types.NamespacedName{Name: serviceAccountName, Namespace: buildRun.Namespace}, serviceAccount)
-			if err != nil {
-				return nil, err
-			}
+		// we default to pipeline/default sa
+		serviceAccount, err = getDefaultNamespaceSA(ctx, client, buildRun)
```

Comment:
We could do that. Note, this logic was there already. Wondering if we can take this one via a separate issue, for simplicity. You tell me.

Reply:
If we validated this synchronously with a webhook validator, it would be less possible to get stuck in this loop (you'd have to create the SA, create the BuildRun, then quickly delete the SA before reconciliation). Synchronous validation would also remove a lot of the logic in this reconcile loop, which I think would make it easier to read and understand as well.

Reply:
Is okay for me.

Reply:
so I created https://github.com/shipwright-io/build/issues/662 for this particular topic. @ImJasonH can you elaborate more on the synchronously validation pls, or maybe do that in 662. I was not able to fully get the idea for the scenario. we are discussing in here.